### PR TITLE
[3006.x] Update 3006 packaging to reduce permissions given to salt user for running salt-master

### LINKED ISSUE #64193: [BUG] 3006.0 Debian and RPM salt-master package creates insecure permissions for salt user
**Description**

The release of 3006.0 included changes in the Debian and RPM packaging to create a user `salt` and config to run the `salt-master` processes under this user (https://github.com/saltstack/salt/issues/64045 and https://github.com/saltstack/salt/pull/64037).

Whilst running the master as a non-root user is an improvement, the benefit is seriously compromised by giving the `salt` user ownership of 

* /etc/salt
* /var/log/salt
* /opt/saltstack/salt/
* /var/cache/salt/
* /var/run/salt

The main issues being:

* Write access to the python installation in `/opt/saltstack/salt` - the `salt-master` process should not be able to (persistently) modify the code it is running and more importantly, the code other salt processes (eg a local minion, `salt-api`, etc) are running as `root`.

* Read/write access to configuration and private data of other salt processes (eg `salt-minion`)

**Setup**

Install 3006.0 from Debian or RPM packages

- [X] onedir packaging

**Expected behavior**

The `salt-master` process:

* Should not be able to modify the content of `/opt/saltstack/salt`
* Should not be able to read private data from other salt processes
* Should not be able to modify the config of other salt processes

**Versions Report**

* 3006.0

**Additional context**

* Ownership of `/opt/saltstack/salt` by `salt` user - it's not clear from  https://github.com/saltstack/salt/pull/64037, but I suspect that this may have been done so that it is possible for the `salt-master` process to create bytecompiled python modules when it runs. It should be possible to create those at package install time and have them owned by root instead.
* Making `/opt/saltstack/salt` owned by root will mean that the MasterMinion won't be able to install python modules. I think that's a corner case/anti-feature, the default should be that the `salt-master` cannot modify the salt install and if users need to do that they can choose to modify ownership locally or run the master as root.
* It shouldn't be necessary for the `salt-master` to be able to write to anything other than `/etc/salt/pki/master` and `/etc/salt/master.d` under `/etc/salt`.
* It shouldn't really be necessary for the `salt-master` process to read any configs for other salt commands/daemons from `/etc/salt`, but it looks like it may currently need to read the minion config (see https://github.com/saltstack/salt/issues/64158)
* A design for handing users and permissions for running salt daemons as non-root users was outlined and accepted in [SEP19](https://github.com/saltstack/salt-enhancement-proposals/blob/master/accepted/0019-master-non-root.md) but this has not been followed in the current implementation.

### PR DESCRIPTION
**Pull request by barneysowood**
### What does this PR do?

Updates Debian and RPM packages to improve isolation offered by running the `salt-master` processes under the `salt` user.

Implementation of running salt-master as a non-root user by default introduced in https://github.com/saltstack/salt/pull/64037, gives ownership of shared salt directories to the `salt` user to mitigate a number of issues with running the `salt-master` as a non-root user. This PR aims to reduce the scope of those permission grants to increase isolation of the `salt-master` processes running under the `salt` user - see https://github.com/saltstack/salt/issues/64193 for more detail.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/64193

### Previous Behavior
The `salt` user that the `salt-master` process runs as was able to:

* Write to `/opt/saltstack/salt` modifying the salt install, including python, shared libs and python modules used by both the salt master and other salt process (eg `salt-api` and `salt-minion`) which run as root, compromising the isolation offered by running the `salt-master` as a non-root user.
* Read and write configs in `/etc/salt` for other salt daemons (that run as root) compromising the isolation offered by running as a non-root user.
* Reading private data from `/etc/salt/pki/minion` and `/var/cache/salt/minion` again compromising isolation.

### New Behavior

Debian and RPM packages will now:

* Keep ownership of `/opt/saltstack/salt` hierarchy as `root:root` to prevent modification by `salt-master` process
* Bytecompile python packages under `/opt/saltstack/salt/lib/` on install so that performance isn't compromised and the bytecompiled modules are owned by root
* Clean up *.pyc files and \_\_pycache__ dirs under `/opt/saltstack/salt/lib` on uninstall
* Under /etc/salt limit ownership to  /etc/salt/pki/master and /etc/salt/master.d
* Under /var/cache/salt and /var/run/salt only give ownership on master directories
* Under /var/log/salt, ensure /var/log/salt/master exists and give ownership of that. Also update logrotate config to create that with correct ownership and perms and install that on debian packages.

This should ensure that a compromised `salt-master` process cannot:

* Modify the salt installation used by it or other salt-daemons
* Modify the config of other salt daemons or tools
* Read pki or other private data *via the local filesystem* for a `salt-minion` running on the same host

### Merge requirements satisfied?

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
No

### CODE REVIEW

**Comment by dwoz** [comment]
This is going to be affected by #64174

**Review by twangboy — DISMISSED** [review]

**Comment by barneysowood** [comment]
As discussed in [#64174](https://github.com/saltstack/salt/pull/64174#issuecomment-1531660628), will wait for that to be merged then re-base this and integrate on top of those changes.
Also need to add some tests.

**Comment by barneysowood** [comment]
Note that issues in https://github.com/saltstack/salt/issues/64219 will affect this.

**Comment by dwoz** [comment]
What is the `saltZ` user?

**Comment by barneysowood** [comment]
> What is the `saltZ` user?

I'm getting a test failure that is throwing an error on chmod for the salt user not existing. That was just a hack to try and work out which chmod call was causing the issue. 

It'd really help if I could get the artifacts from test runs to get the pytest logs - they don't seem to be accessible from the workflows page - is there some other way to get access to them?

**Comment by dmurphy18** [comment]
@barneysowood Wondering if you have a timeframe for when this will come out of draft, itching to review, but don't want to, till it is out of draft.

**Comment by barneysowood** [comment]
> @barneysowood Wondering if you have a timeframe for when this will come out of draft, itching to review, but don't want to, till it is out of draft.

@dmurphy18 - apologies, I've been out of action for a while. I'm planning to pick this up later this week, IIRC I've just got some issues with tests to fix. If I can get those sorted quickly I will do so, if not I may take it out of draft for review anyway whilst I work on those.

**Comment by Ch3LL** [comment]
@barneysowood this is still in draft mode. We hope to release the next bug fix really soon, mid August. Is this close to being ready?

**Comment by barneysowood** [comment]
> @barneysowood this is still in draft mode. We hope to release the next bug fix really soon, mid August. Is this close to being ready?

@Ch3LL - I think so. I'm hoping to have to one remaining issue that blocking tests passing fixed today. I'll let you know how that goes.

**Comment by barneysowood** [comment]
re-run all

**Comment by barneysowood** [comment]
Seeing a strange issue with tests for this. All tests now passing apart from "Test (upgrade, 3006.1)" for Debian 10/11 and Ubuntu 20.04/22.04 - see https://github.com/saltstack/salt/actions/runs/5801918905/job/15729830349?pr=64194 for an example.

What's failing specifically is `pkg/tests/integration/test_pip.py::test_pip_non_root` - 

```                   # Now, we should still not be able to install as non-root
                   ret = subprocess.run(
                       install_salt.binary_paths["pip"] + ["install", "pep8"],
                       preexec_fn=demote(test_account.uid, test_account.gid),
                       env=test_account.env,
                       stdout=subprocess.PIPE,
                       stderr=subprocess.PIPE,
                       check=False,
                       universal_newlines=True,
                   )
           >       assert ret.returncode != 0, ret.stderr
           E       AssertionError:
```

So that's testing that running `/opt/saltstack/salt/salt-pip install pep8` as a non-root user should fail. I can't recreate this when testing locally - taking the built artifacts and applying them to a VM with 3006.1 installed doesn't show the same behaviour.

Have also tested using the `ghcr.io/saltstack/salt-ci-containers/packaging:debian-11` container locally running `nox -e "test-pkgs-onedir" -- upgrade --prev-version 3006.1` and that doesn't show the failure.

Slightly stumped at this point - I suspect some interaction with pytest-salt-factories and the specific setup with CI, but that's really hard to debug.

@Ch3LL, @s0undt3ch or @dmurphy18 - any ideas on where to go next?

As this is the one remaining issue on this PR from my point of view, going to take it out of draft..

File: pkg/debian/salt-master.postinst
Comment (s0undt3ch) [inline pkg/debian/salt-master.postinst:11]:
```suggestion
    chown -R salt:salt /etc/salt/pki/master /etc/salt/master.d /var/log/salt/master /var/log/salt/key /var/cache/salt/master /var/run/salt/master
```
Should `/etc/salt/minion.d` really be above?

File: pkg/debian/salt-master.preinst
Comment (s0undt3ch) [inline pkg/debian/salt-master.preinst:7]:
```suggestion
    PY_VER=$(/opt/saltstack/salt/bin/python3 -c "import sys; sys.stdout.write('{}.{}'.format(*sys.version_info)); sys.stdout.flush();")
```

File: pkg/rpm/salt.spec
Comment (s0undt3ch) [inline pkg/rpm/salt.spec:414]:
```suggestion
PY_VER=$(/opt/saltstack/salt/bin/python3 -c "import sys; sys.stdout.write('{}.{}'.format(*sys.version_info)); sys.stdout.flush();")
```

File: pkg/rpm/salt.spec
Comment (s0undt3ch) [inline pkg/rpm/salt.spec:493]:
```suggestion
chown -R %{_SALT_USER}:%{_SALT_GROUP} /etc/salt/cloud.deploy.d /var/log/salt/cloud /opt/saltstack/salt/lib/python3.10/site-packages/salt/cloud/deploy
```
Is there a way not to have `python3.10` hardcoded?

File: pkg/tests/integration/test_salt_user.py
Comment (s0undt3ch) [inline pkg/tests/integration/test_salt_user.py:29]:
```suggestion
    return [
        "/etc/salt",
        "/var/cache/salt",
        "/var/log/salt",
        "/var/run/salt",
        "/opt/saltstack/salt",
    ]
```

File: pkg/tests/integration/test_salt_user.py
Comment (s0undt3ch) [inline pkg/tests/integration/test_salt_user.py:114]:
```suggestion
        "/opt/saltstack/salt/lib/python3.10/site-packages/salt/cloud/deploy",
```
Don't hardcode the python version. Why did you need to change this line?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: pkg/tests/integration/test_salt_user.py
Comment (barneysowood) [reply pkg/tests/integration/test_salt_user.py:114]: Just accidentally stomped on that - reverted.

File: pkg/debian/salt-master.preinst
Comment (barneysowood) [reply pkg/debian/salt-master.preinst]: Fixed

File: pkg/rpm/salt.spec
Comment (barneysowood) [reply pkg/rpm/salt.spec]: Fixed

File: pkg/tests/integration/test_salt_user.py
Comment (barneysowood) [reply pkg/tests/integration/test_salt_user.py:115]: I made the change in https://github.com/saltstack/salt/pull/64194/commits/1b411c7fb5cb4d07ddb24dae8b6c18953ea4d987 a couple of days before your comment.

Comment (s0undt3ch) [reply pkg/tests/integration/test_salt_user.py:115]: > I made the change in [1b411c7](https://github.com/saltstack/salt/commit/1b411c7fb5cb4d07ddb24dae8b6c18953ea4d987) a couple of days before your comment.

There's one more hardcoded reference, the one being commented on

Comment (barneysowood) [reply pkg/tests/integration/test_salt_user.py]: @s0undt3ch - apologies, you're correct. I've actually removed test_salt_cloud_dirs() as that is now covered by  test_pkg_paths()

**Review by s0undt3ch — DISMISSED** [review]

**Review by dmurphy18 — DISMISSED** [review]
using perl for regex, hurts for python :)

**Comment by Ch3LL** [comment]
@barneysowood these test failures look like they might be related. Note: We are trying to get everything merged in by this Wednesday for the next bug fix release.

**Comment by barneysowood** [comment]
> @barneysowood these test failures look like they might be related. Note: We are trying to get everything merged in by this Wednesday for the next bug fix release.

@Ch3LL  - I've been trying to debug them, making some progress, but it's rather slow. When you say "these test failures look like they might be related" - what exactly do you mean? Related to what?

**Comment by Ch3LL** [comment]
Related to the code changes in this PR, especially since I haven't seen this test failure in any other PRs. I will update the branch just to be 100% certain.

I have also asked @MKLeb to help you take a look at the failures, since we would like to get this in for the bug fix release. He said he can dive into it tomorrow to take a look.

**Comment by barneysowood** [comment]
> Related to the code changes in this PR, especially since I haven't seen this test failure in any other PRs. I will update the branch just to be 100% certain.
> 
> I have also asked @MKLeb to help you take a look at the failures, since we would like to get this in for the bug fix release. He said he can dive into it tomorrow to take a look.

@Ch3LL - ok thanks.

@MKLeb - would very much appreciate some help, pretty stumped with why `pkg/tests/integration/test_pip.py::test_pip_non_root` is failing on the 3006.2 upgrade. 

I created a branch with just some changes to that test to dump debug info and a PR to run the test suite.. results are weird.

Pulled the logs from this - https://github.com/saltstack/salt/actions/runs/5876536569/job/15938431199?pr=64947

The issue is that somehow running `salt-pip install` as the saltdev user works - it creates a dir in `/opt/saltstack/salt`.. I added some debugging to dump out `ls -l` for `/opt/saltstack`, `/opt/saltstack/salt` and `/opt/saltstack/salt/extras-3.10` before the run of `salt-pip install -h`, then before run of `salt-pip install pep8` and after that.

1. Output of `id saltdev`

```
uid=1002(saltdev) gid=1002(saltdev) groups=1002(saltdev)
```

2. Output of  `/bin/ls -l /opt/saltstack /opt/saltstack/salt /opt/saltstack/salt/extras-3.10` at start of test

```
/opt/saltstack:
total 4
drwxrwxr-x 6 root root 4096 Aug 16 10:42 salt

/opt/saltstack/salt:
total 92
drwxr-xr-x  2 root root  4096 Aug 16 10:40 bin
drwxr-xr-x 12 root root  4096 Aug 16 10:40 include
drwxr-xr-x  6 root root 20480 Aug 16 10:40 lib
-rwxr-xr-x  1 root root   295 Aug 16 08:37 salt
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-api
-rwxr-xr-x  1 root root   295 Aug 16 08:37 salt-call
-rwxr-xr-x  1 root root   297 Aug 16 08:37 salt-cloud
-rwxr-xr-x  1 root root   291 Aug 16 08:37 salt-cp
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-key
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-master
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-minion
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-pip
-rwxr-xr-x  1 root root   297 Aug 16 08:37 salt-proxy
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-run
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-ssh
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-syndic
drwxr-xr-x  3 root root  4096 Aug 16 10:39 share
-rwxr-xr-x  1 root root   293 Aug 16 08:37 spm
-rwxr-xr-x  1 root root   282 Aug 16 08:37 wheel

 /bin/ls: cannot access '/opt/saltstack/salt/extras-3.10': No such file or directory
```

3. Output of  `/bin/ls -l /opt/saltstack /opt/saltstack/salt /opt/saltstack/salt/extras-3.10` after `salt-pip install -h` but before `salt-pip install pep8`

```
/opt/saltstack:
total 4
drwxrwxr-x 6 root root 4096 Aug 16 10:42 salt

/opt/saltstack/salt:
total 92
drwxr-xr-x  2 root root  4096 Aug 16 10:40 bin
drwxr-xr-x 12 root root  4096 Aug 16 10:40 include
drwxr-xr-x  6 root root 20480 Aug 16 10:40 lib
-rwxr-xr-x  1 root root   295 Aug 16 08:37 salt
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-api
-rwxr-xr-x  1 root root   295 Aug 16 08:37 salt-call
-rwxr-xr-x  1 root root   297 Aug 16 08:37 salt-cloud
-rwxr-xr-x  1 root root   291 Aug 16 08:37 salt-cp
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-key
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-master
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-minion
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-pip
-rwxr-xr-x  1 root root   297 Aug 16 08:37 salt-proxy
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-run
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-ssh
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-syndic
drwxr-xr-x  3 root root  4096 Aug 16 10:39 share
-rwxr-xr-x  1 root root   293 Aug 16 08:37 spm
-rwxr-xr-x  1 root root   282 Aug 16 08:37 wheel

/bin/ls: cannot access '/opt/saltstack/salt/extras-3.10

```

4. Output of  `/bin/ls -l /opt/saltstack /opt/saltstack/salt /opt/saltstack/salt/extras-3.10` after `salt-pip install pep8`

```
/opt/saltstack:
total 4
drwxrwxr-x 7 root root 4096 Aug 16 10:42 salt

/opt/saltstack/salt:
total 96
drwxr-xr-x  2 root    root     4096 Aug 16 10:40 bin
drwxr-xr-x  5 saltdev saltdev  4096 Aug 16 10:42 extras-3.10
drwxr-xr-x 12 root    root     4096 Aug 16 10:40 include
drwxr-xr-x  6 root    root    20480 Aug 16 10:40 lib
-rwxr-xr-x  1 root    root      295 Aug 16 08:37 salt
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-api
-rwxr-xr-x  1 root    root      295 Aug 16 08:37 salt-call
-rwxr-xr-x  1 root    root      297 Aug 16 08:37 salt-cloud
-rwxr-xr-x  1 root    root      291 Aug 16 08:37 salt-cp
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-key
-rwxr-xr-x  1 root    root      299 Aug 16 08:37 salt-master
-rwxr-xr-x  1 root    root      299 Aug 16 08:37 salt-minion
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-pip
-rwxr-xr-x  1 root    root      297 Aug 16 08:37 salt-proxy
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-run
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-ssh
-rwxr-xr-x  1 root    root      299 Aug 16 08:37 salt-syndic
drwxr-xr-x  3 root    root     4096 Aug 16 10:39 share
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 spm
-rwxr-xr-x  1 root    root      282 Aug 16 08:37 wheel

/opt/saltstack/salt/extras-3.10:
total 96
drwxr-xr-x 2 saltdev saltdev  4096 Aug 16 10:42 bin
drwxr-xr-x 2 saltdev saltdev  4096 Aug 16 10:42 pep8-1.7.1.dist-info
-rw-r--r-- 1 saltdev saltdev 82878 Aug 16 10:42 pep8.py
drwxr-xr-x 2 saltdev saltdev  4096 Aug 16 10:42 __pycache__

```

[full log](https://github.com/saltstack/salt/files/12358635/runtests-upgrade.log)

So, somehow, despite `/opt/saltstack/salt` being `0775 root:root`, the saltdev user is able to create a `extras-3.10` in it. Maybe I'm missing something obvious here - I have been staring at it a long time.

**Comment by Ch3LL** [comment]
@barneysowood looks like there is a merge conflict that needs to be resolved here now. I'm guessing it was caused by this: https://github.com/saltstack/salt/pull/64378

**Comment by barneysowood** [comment]
> @barneysowood looks like there is a merge conflict that needs to be resolved here now. I'm guessing it was caused by this: #64378

Resolved that.

**Comment by barneysowood** [comment]
Created #65001 - which can be merged once this and #64235 have been merged to 3006.x

**Comment by dwoz** [comment]
Have we tested upgrading from a 3005.x package?

**Comment by dwoz** [comment]
> Have we tested upgrading from a 3005.x package?

@MKLeb Let me know we have tests that test the upgrades from 3005.x.

File: pkg/debian/salt-common.preinst
Comment (MKLeb) [inline pkg/debian/salt-common.preinst:34]: Hi @barneysowood, are you able to elaborate on why you are removing this piece?

**Comment by barneysowood** [comment]
> > Have we tested upgrading from a 3005.x package?
> 
> @MKLeb Let me know we have tests that test the upgrades from 3005.x.

We do and I've also manually tested that on Ubuntu and CentOS.

File: pkg/debian/salt-common.preinst
Comment (barneysowood) [reply pkg/debian/salt-common.preinst:34]: Hi @MKLeb - because the `salt` owner that the salt-master process runs as shouldn't be able to write to /opt/saltstack/salt - this would allow it modify the whole salt install...

**Review by Ch3LL — APPROVED** [review]

**Review by MKLeb — APPROVED** [review]

**Review by dmurphy18 — APPROVED** [review]

**Review by dwoz — APPROVED** [review]

**Pull request merged**
